scipy/optimize/tests/test_optimize.py

```diff
@@ -44,6 +44,135 @@ def der_logit(x):
     assert_(r > 1e-7)
 
 
+class TestApproxJacobian(object):
+    def fun_scalar_scalar(self, x):
+        return np.sinh(x)
+
+    def jac_scalar_scalar(self, x):
+        return np.cosh(x)
+
+    def fun_scalar_vector(self, x):
+        return np.array([x**2, np.tan(x), np.exp(x)])
+
+    def jac_scalar_vector(self, x):
+        return np.array([2 * x, np.cos(x) ** -2, np.exp(x)])
+
+    def fun_vector_scalar(self, x):
+        return np.sin(x[0] * x[1]) * np.log(x[0])
+
+    def jac_vector_scalar(self, x):
+        return np.array([
+            x[1] * np.cos(x[0] * x[1]) * np.log(x[0]) +
+            np.sin(x[0] * x[1]) / x[0],
+            x[0] * np.cos(x[0] * x[1]) * np.log(x[0])
+        ])
+
+    def fun_vector_vector(self, x):
+        return np.array([
+            x[0] * np.sin(x[1]),
+            x[1] * np.cos(x[0]),
+            x[0] ** 3 * x[1] ** -0.5
```

Comment:
(really minor nitpick) Don't use extra spaces for `x**n`. Compare what you wrote and `x[0]**3 * x[1]**(-0.5)` (brackets are optional).